storage: assert iterator invariants

erikgrinaker opened the pull request:
Resolves #86651.

**storage: remove `Valid()` call in `pebbleIterator.UnsafeKey()`**

This patch removes an unnecessary `Valid()` call in
`pebbleIterator.UnsafeKey()`. The `SimpleMVCCIterator` API contract
already mandates that the caller must call `Valid()` before calling
`UnsafeKey()`.

This is done in preparation for iterator invariant checks. These will be
performed in `Valid()` and this call would cause infinite recursion.
This additionally gives a nice performance boost (using the
`SSTIterator` benchmark since `MVCCScan` does not call `UnsafeKey`):

```
name                                                   old time/op  new time/op  delta
SSTIterator/keys=10000/variant=pebble/verify=false-24  1.08ms ± 0%  1.05ms ± 0%  -2.51%  (p=0.000 n=10+8)
```

Release note: None
  
**storage: add `pebbleIterator` invariant assertions**

This patch adds invariant checks for `pebbleIterator`. General API
invariant checks for the `SimpleMVCCIterator` and `MVCCIterator`
interfaces are also added, and applied to `pebbleIterator`.

Release note: None
  
**storage: add `intentInterleavingIter` invariant assertions**

This patch adds invariant assertions for `intentInterleavingIter` using
a similar pattern as `pointSynthesizingIter`: by calling
`assertInvariants()` under race during `Valid()`. This allows specifying
all internal invariants in a simple-to-understand form and asserting
them all after every state transition.

Existing race assertions have been migrated.

Release note: None
  
**storage: fix `pointSynthesizingIter.Key()` and `UnsafeRawKey()`**

`pointSynthesizingIter.Key()` could return a different key than
`UnsafeKey()`. Additionally, `UnsafeRawKey()` would omit the key's
timestamp. Luckily, these methods have no current callers.

Release note: None

**storage: add `pointSynthesizingIter` invariant assertions**

Release note: None

**storage: add `MVCCIncrementalIterator` invariant assertions**

Release note: None
  
**storage: add `ReadAsOfIterator` invariant assertions**

Release note: None

erikgrinaker committed:
storage: add `intentInterleavingIter` invariant assertions

This patch adds invariant assertions for `intentInterleavingIter` using
a similar pattern as `pointSynthesizingIter`: by calling
`assertInvariants()` under race during `Valid()`. This allows specifying
all internal invariants in a simple-to-understand form and asserting
them all after every state transition.

Existing race assertions have been migrated.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87134)
<!-- Reviewable:end -->

tbg reviewed (dismissed):
Did not review but +1 on the pattern
Req changes so that you can re-request when ready

sumeerbhola reviewed (commented):
+1 on centralizing the assertions. If we ever do https://github.com/cockroachdb/cockroach/issues/66869#issuecomment-1177729588, this will help there too.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87134#-:-NBJ9F6Q5CAXzNmCw7i0:b-m1zi5c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->

erikgrinaker committed:
storage: add `pebbleIterator` invariant assertions

This patch adds invariant checks for `pebbleIterator`. General API
invariant checks for the `SimpleMVCCIterator` and `MVCCIterator`
interfaces are also added, and applied to `pebbleIterator`.

Release note: None

erikgrinaker committed:
storage: remove `Valid()` call in `pebbleIterator.UnsafeKey()`

This patch removes an unnecessary `Valid()` call in
`pebbleIterator.UnsafeKey()`. The `SimpleMVCCIterator` API contract
already mandates that the caller must call `Valid()` before calling
`UnsafeKey()`.

This is done in preparation for iterator invariant checks. These will be
performed in `Valid()` and this call would cause infinite recursion.
This additionally gives a nice performance boost (using the
`SSTIterator` benchmark since `MVCCScan` does not call `UnsafeKey`):

```
name                                                   old time/op  new time/op  delta
SSTIterator/keys=10000/variant=pebble/verify=false-24  1.08ms ± 0%  1.05ms ± 0%  -2.51%  (p=0.000 n=10+8)
```

Release note: None

erikgrinaker committed:
storage: fix `pointSynthesizingIter.Key()` and `UnsafeRawKey()`

`pointSynthesizingIter.Key()` could return a different key than
`UnsafeKey()`. Additionally, `UnsafeRawKey()` would omit the key's
timestamp. Luckily, these methods have no current callers.

Release note: None

erikgrinaker committed:
storage: add `pointSynthesizingIter` invariant assertions

Release note: None

erikgrinaker committed:
storage: assert `MVCCIncrementalIterator` invariant assertions

Release note: None

erikgrinaker committed:
storage: add `ReadAsOfIterator` invariant assertions

Release note: None

erikgrinaker commented:
This should be ready to review now. Found a couple of bugs, and a bonus performance optimization as well.

jbowens reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 8 of 8 files at r9, 2 of 2 files at r10, 4 of 4 files at r11, 1 of 1 files at r12, 1 of 1 files at r13, 2 of 2 files at r14, 1 of 1 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87134#-:-NBwUawq2AEOScXQg5pe:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->

erikgrinaker commented:
TFTR!

bors r=jbowens

erikgrinaker commented:
bors r=jbowens

erikgrinaker commented:
bors retry
